pkg/splunk/enterprise/upgrade.go

```diff
@@ -0,0 +1,315 @@
+			scopedLog.Error(err, "Unable to get licenseManager current image")
+			return false, err
+		}
+		// if license manager status is ready and CR spec and current license manager image are not same
+		// then return with error
+		if licenseManager.Status.Phase != enterpriseApi.PhaseReady || lmImage != spec.Image {
+			return false, fmt.Errorf("license manager is not ready or license manager current image is different than CR image")
+		}
+		goto ClusterManager
+	}
+ClusterManager:
+	if cr.GroupVersionKind().Kind == "ClusterManager" {
+
+		licenseManagerRef := spec.LicenseManagerRef
+		if licenseManagerRef.Name == "" {
+			return true, nil
+		}
+		namespacedName := types.NamespacedName {
+			Namespace: cr.GetNamespace(),
+			Name:      GetSplunkStatefulsetName(SplunkClusterManager, cr.GetName()),
+		}
+
+		// check if the stateful set is created at this instance
+		statefulSet := &appsv1.StatefulSet{}
+		err := c.Get(ctx, namespacedName, statefulSet)
+		if err != nil {
+			if k8serrors.IsNotFound(err) {
+				return true, nil
+			}
+			return false, nil
+		}
+		return true, nil
+	} else {
+		// check if a cluster manager reference is added to custom resource
+		clusterManagerRef := spec.ClusterManagerRef
+		if clusterManagerRef.Name == "" {
+			// if ref is not defined go to monitoring console step
+			goto MonitoringConsole
+		}
+
+		namespacedName := types.NamespacedName{Namespace: cr.GetNamespace(), Name: clusterManagerRef.Name}
+		clusterManager := &enterpriseApi.ClusterManager{}
+
+		// get the cluster manager referred in custom resource
+		err := c.Get(ctx, namespacedName, clusterManager)
+		if err != nil {
+			eventPublisher.Warning(ctx, "UpgradePathValidation", fmt.Sprintf("Could not find the Cluster Manager. Reason %v", err))
+			scopedLog.Error(err, "Unable to get clusterManager")
+			goto MonitoringConsole
+		}
+
+		/// get the cluster manager image referred in custom resource
+		cmImage, err := getCurrentImage(ctx, c, clusterManager, SplunkClusterManager)
+		if err != nil {
+			eventPublisher.Warning(ctx, "UpgradePathValidation", fmt.Sprintf("Could not get the Cluster Manager Image. Reason %v", err))
+			scopedLog.Error(err, "Unable to get clusterManager current image")
+			return false, err
+		}
+
+		// check if an image upgrade is happening and whether CM has finished updating yet, return false to stop
+		// further reconcile operations on custom resource until CM is ready
+		if clusterManager.Status.Phase != enterpriseApi.PhaseReady || cmImage != spec.Image {
+			return false, nil
+		}
+		goto MonitoringConsole
+	}
+MonitoringConsole:
+	if cr.GroupVersionKind().Kind == "MonitoringConsole" {
+
+		namespacedName := types.NamespacedName{
+			Namespace: cr.GetNamespace(),
+			Name:      GetSplunkStatefulsetName(SplunkMonitoringConsole, cr.GetName()),
+		}
+
+		// check if the stateful set is created at this instance
+		statefulSet := &appsv1.StatefulSet{}
+		err := c.Get(ctx, namespacedName, statefulSet)
```

Comment:
Same as above and below instances

Reply:
well, possible that once in a while API server fails as too many calls to API server